lib/ansible/modules/cloud/amazon/s3_facts.py

```diff
@@ -0,0 +1,155 @@
+
+# Lists all s3 buckets
+- s3_facts:
+'''
+
+RETURN = '''
+buckets:
+  description: List of buckets
+  returned: always
+  type: list
+'''
+
+import json
+import traceback
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ec2 import get_aws_connection_info, ec2_argument_spec, AnsibleAWSError, boto3_conn, \
+    camel_dict_to_snake_dict, connect_to_aws
+
+try:
+    import boto.ec2
+    from boto.exception import BotoServerError, S3ResponseError
+
+    HAS_BOTO = True
+except ImportError:
+    HAS_BOTO = False
+
+try:
+    import boto3
+    from botocore.exceptions import ClientError, NoCredentialsError
+
+    HAS_BOTO3 = True
+except ImportError:
+    HAS_BOTO3 = False
+
+
+class S3ServiceManager:
+    '''
+    Service manager for S3, supporting both boto and boto3
+    '''
+
+    def __init__(self, module):
+        '''
+        Bootstrap module and connection
+        :param module:
+        '''
+
+        self.module = module
+
+        # Set up connection
+        region, ec2_url, aws_connect_params = get_aws_connection_info(self.module, boto3=HAS_BOTO3)
+
+        if region:
+            if HAS_BOTO3:
+                try:
+                    self.connection = boto3_conn(self.module, conn_type='client', resource='s3', region=region,
+                                                 endpoint=ec2_url,
+                                                 **aws_connect_params)
+                except ClientError as e:
+                    self.module.fail_json(msg=e.message, exception=traceback.format_exc(),
+                                          **camel_dict_to_snake_dict(e.response))
+            if HAS_BOTO:
+                try:
+                    self.connection = connect_to_aws(boto.s3, region, **aws_connect_params)
+                except (boto.exception.NoAuthHandlerFound, AnsibleAWSError) as e:
+                    self.module.fail_json(msg=str(e))
+        else:
+            self.module.fail_json(msg="AWS region must be specified (like: us-east-1)")
+
+    def get_bucket_list(self):
+        '''
+        Get bucket list
+        :return:
+        '''
+
+        try:
+            bucket_list = self.connection.get_all_buckets()
+        except S3ResponseError as e:
+            self.module.fail_json(msg=e.message)
+
+        # If the list is not empty, repackage so the names end up in a json based string
+        if bucket_list is not None:
+            bucket_list_json = json.dumps([bucket.name for bucket in bucket_list])
+        else:
+            bucket_list_json = json.dumps([])
+
+        return bucket_list_json
+
+
+def main():
+    '''
+    Run module
+    :return:
+    '''
+
+    # Including ec2 argument spec, no support for check mode as it will always return new results
+    module = AnsibleModule(argument_spec=ec2_argument_spec(), supports_check_mode=False)
+
+    if not HAS_BOTO:
+        module.fail_json(msg='boto required for this module')
+
+    # Set defaults
+    result = {'s3': {}}
```

Comment:
It looks like the return values are `{'s3': {'buckets': [.......]}}` which isn't what's in the docs.